pkg/runtime/grpc/grpc-runtime.go

```diff
@@ -184,17 +227,79 @@ nodesLoop:
 	return res, nil
 }
 
+func (r *Runtime) getTargets(ctx context.Context, params *params.Params) ([]target, error) {
+	switch r.connectionMode {
+	case ConnectionModeKubernetesProxy:
+		// Get nodes to run on
+		nodes := params.Get(ParamNode).AsStringSlice()
+		pods, err := getGadgetPods(ctx, nodes)
+		if err != nil {
+			return nil, fmt.Errorf("get gadget pods: %w", err)
+		}
+		if len(pods) == 0 {
+			return nil, fmt.Errorf("get gadget pods: Inspektor Gadget is not running on the requested node(s): %v", nodes) //nolint:all
+		}
+		return pods, nil
+	case ConnectionModeDirect:
+		inTargets := params.Get(ParamRemoteAddress).AsStringSlice()
+		targets := make([]target, 0)
+		for _, t := range inTargets {
+			purl, err := url.Parse(t)
+			if err != nil {
+				return nil, fmt.Errorf("invalid remote address %q: %w", t, err)
+			}
+			targets = append(targets, target{
+				name: t,
+				node: purl.Hostname(),
+			})
+		}
+		return targets, nil
+	}
+	return nil, fmt.Errorf("unsupported connection mode")
+}
+
 func (r *Runtime) RunGadget(gadgetCtx runtime.GadgetContext) (runtime.CombinedGadgetResult, error) {
-	// Get nodes to run on
-	nodes := gadgetCtx.RuntimeParams().Get(ParamNode).AsStringSlice()
-	pods, err := getGadgetPods(gadgetCtx.Context(), nodes)
+	paramMap := make(map[string]string)
+	gadgets.ParamsToMap(
+		paramMap,
+		gadgetCtx.GadgetParams(),
+		gadgetCtx.RuntimeParams(),
+		gadgetCtx.OperatorsParamCollection(),
+	)
+
+	gadgetCtx.Logger().Debugf("Params")
+	for k, v := range paramMap {
+		gadgetCtx.Logger().Debugf("- %s: %q", k, v)
+	}
+
+	targets, err := r.getTargets(gadgetCtx.Context(), gadgetCtx.RuntimeParams())
+	if err != nil {
+		return nil, fmt.Errorf("getting target nodes: %w", err)
+	}
+	return r.runGadgetOnTargets(gadgetCtx, paramMap, targets)
+}
+
+func (r *Runtime) getClientFromRandomTarget(ctx context.Context, runtimeParams *params.Params) (api.GadgetManagerClient, error) {
+	targets, err := r.getTargets(ctx, runtimeParams)
 	if err != nil {
-		return nil, fmt.Errorf("get gadget pods: %w", err)
+		return nil, err
 	}
-	if len(pods) == 0 {
-		return nil, fmt.Errorf("get gadget pods: Inspektor Gadget is not running on the requested node(s): %v", nodes) //nolint:all
+	if len(targets) == 0 {
+		return nil, fmt.Errorf("no valid targets")
 	}
+	conn, err := r.dialContext(ctx, targets[0])
+	if err != nil {
+		return nil, err
+	}
+	client := api.NewGadgetManagerClient(conn)
+	return client, nil
```

Comment:
super minor nit: Avoid the temp variable and just `return api.NewGadgetManagerClient(conn), nil`?